base/applications/cleanmgr/util.c

```diff
@@ -0,0 +1,1179 @@
+    FileStruct.fAnyOperationsAborted = FALSE;
+    FileStruct.pFrom = TempPath;
+
+    SHFileOperationW(&FileStruct);
+    free(Path);
+}
+
+uint64_t DirSizeFunc(PWCHAR TargetDir)
+{
+    WIN32_FIND_DATAW DataStruct;
+    ZeroMemory(&DataStruct, sizeof(DataStruct));
+    LARGE_INTEGER sz;
+    ZeroMemory(&sz, sizeof(sz));
+    HANDLE HandleDir = NULL;
+    uint64_t Size = 0;
+    WCHAR TargetedDir[2048] = { 0 };
+    WCHAR ReTargetDir[2048] = { 0 };
+    DWORD RetError = 0;
+    
+    StringCchPrintfW(TargetedDir, _countof(TargetedDir), L"%s%s", TargetDir, L"\\*");
+    HandleDir = FindFirstFileW(TargetedDir, &DataStruct);
+    RetError = GetLastError();
+    
+    if (RetError != ERROR_NO_MORE_FILES && HandleDir == INVALID_HANDLE_VALUE)
+    {
+        if (RetError == ERROR_PATH_NOT_FOUND)
+        {
+            return 0;
+        }
+
+        else
+        {
+            MessageBoxW(NULL, L"FindFirstFileW() failed!", L"Error", MB_OK | MB_ICONERROR);
+            return 1;
+        }
+    }
+
+    do
+    {
+        if (wcscmp(DataStruct.cFileName, L".") != 0 && wcscmp(DataStruct.cFileName, L"..") != 0)
+        {
+            if (DataStruct.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY)
+            {
+                StringCchPrintfW(ReTargetDir, _countof(ReTargetDir), L"%s\\%s", TargetDir, DataStruct.cFileName);
+                Size = DirSizeFunc(ReTargetDir);
+            }
+
+            else
+            {
+                sz.LowPart = DataStruct.nFileSizeLow;
+                sz.HighPart = DataStruct.nFileSizeHigh;
+                Size += sz.QuadPart;
+            }
+        }
+    }
+
+    while (FindNextFileW(HandleDir, &DataStruct));
+
+    FindClose(HandleDir);
+
+    return Size;
+}
+
+double SetOptimalSize(uint64_t Size)
+{
+    if ((double)Size >= (GB))
+    {
+        return (double)Size / (GB);
+    }
+
+    else if ((double)Size >= (MB))
+    {
+        return (double)Size / (MB);
+    }
+
+    else if ((double)Size >= (KB))
+    {
+        return (double)Size / (KB);
+    }
+
+    return (double)Size;
+}
+
+void AddItem(HWND hList, PWCHAR String, PWCHAR SubString, int iIndex)
+{
+    LVITEMW lvI;
+    ZeroMemory(&lvI, sizeof(lvI));
+    
+    lvI.mask = LVIF_TEXT | LVIF_IMAGE;
+    lvI.cchTextMax = 256;
+    lvI.iItem = SendMessage(hList, LVM_GETITEMCOUNT, 0, 0);
+    lvI.iSubItem = 0;
+    lvI.iImage = iIndex;
+    lvI.pszText = String;
+    ListView_InsertItem(hList, &lvI);
+    lvI.iSubItem = 1;
+    lvI.pszText = SubString;    
+    ListView_SetItem(hList, &lvI);
+}
+
+BOOL CreateImageLists(HWND hList)
+{
+    HICON hbmIcon;
+    HIMAGELIST hSmall;
+    hSmall = ImageList_Create(16, 16, ILC_COLOR16 | ILC_MASK, 3, 3);
+
+    if(hSmall == NULL)
+        return FALSE;
+    
+    for (int index = 0; index < 3; index++)
+    {
+        hbmIcon = LoadIconW(GetModuleHandleW(NULL), MAKEINTRESOURCE(IDI_BLANK + index));
+        ImageList_AddIcon(hSmall, hbmIcon);
+        DestroyIcon(hbmIcon);
```

Comment:
https://docs.microsoft.com/en-us/windows/win32/api/commctrl/nf-commctrl-imagelist_addicon
> Because the system does not save hicon, you can destroy it after the macro returns if the icon or cursor was created by the CreateIcon function. You do not need to destroy hicon if it was loaded by the LoadIcon function; the system automatically frees an icon resource when it is no longer needed.